cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller.go

```diff
@@ -428,190 +425,128 @@ func getAPIGroupPreferredVersion(client discovery.DiscoveryInterface, APIGroup s
-			continue
+		_, found, err = unstructured.NestedFieldCopy(machine.UnstructuredContent(), "status", "nodeRef")
+		if err != nil {
+			return nil, err
 		}
 
-		if machine.Status.NodeRef.Kind != "Node" {
-			klog.Errorf("Status.NodeRef of machine %q does not reference a node (rather %q)", machine.Name, machine.Status.NodeRef.Kind)
+		if !found {
+			klog.V(4).Infof("Status.NodeRef of machine %q is currently nil", machine.GetName())
 			continue
 		}
 
-		node, err := c.findNodeByNodeName(machine.Status.NodeRef.Name)
+		nodeRefKind, found, err := unstructured.NestedString(machine.UnstructuredContent(), "status", "nodeRef", "kind")
 		if err != nil {
-			return nil, fmt.Errorf("unknown node %q", machine.Status.NodeRef.Name)
+			return nil, err
 		}
 
-		if node != nil {
-			providerIDs = append(providerIDs, node.Spec.ProviderID)
+		if found && nodeRefKind != "Node" {
+			klog.Errorf("Status.NodeRef of machine %q does not reference a node (rather %q)", machine.GetName(), nodeRefKind)
+			continue
 		}
-	}
 
-	klog.V(4).Infof("nodegroup %s has nodes %v", machineSet.Name, providerIDs)
-	return providerIDs, nil
-}
-
-func (c *machineController) filterAllMachineSets(f machineSetFilterFunc) error {
-	return c.filterMachineSets(metav1.NamespaceAll, f)
-}
-
-func (c *machineController) filterMachineSets(namespace string, f machineSetFilterFunc) error {
-	machineSets, err := c.listMachineSets(namespace, labels.Everything())
-	if err != nil {
-		return nil
-	}
-	for _, machineSet := range machineSets {
-		if err := f(machineSet); err != nil {
-			return err
+		nodeRefName, found, err := unstructured.NestedString(machine.UnstructuredContent(), "status", "nodeRef", "name")
+		if err != nil {
+			return nil, err
 		}
-	}
-	return nil
-}
 
-func (c *machineController) machineSetNodeGroups() ([]*nodegroup, error) {
-	var nodegroups []*nodegroup
+		if found {
+			node, err := c.findNodeByNodeName(nodeRefName)
+			if err != nil {
+				return nil, fmt.Errorf("unknown node %q", nodeRefName)
+			}
 
-	if err := c.filterAllMachineSets(func(machineSet *MachineSet) error {
-		if machineSetHasMachineDeploymentOwnerRef(machineSet) {
-			return nil
-		}
-		ng, err := newNodegroupFromMachineSet(c, machineSet)
-		if err != nil {
-			return err
-		}
-		if ng.MaxSize()-ng.MinSize() > 0 && pointer.Int32PtrDerefOr(machineSet.Spec.Replicas, 0) > 0 {
-			nodegroups = append(nodegroups, ng)
+			if node != nil {
+				providerIDs = append(providerIDs, node.Spec.ProviderID)
+			}
 		}
-		return nil
-	}); err != nil {
-		return nil, err
 	}
 
-	return nodegroups, nil
+	klog.V(4).Infof("nodegroup %s has nodes %v", scalableResource.GetName(), providerIDs)
+
+	return providerIDs, nil
 }
 
-func (c *machineController) machineDeploymentNodeGroups() ([]*nodegroup, error) {
-	machineDeployments, err := c.listMachineDeployments(metav1.NamespaceAll, labels.Everything())
+func (c *machineController) nodeGroups() ([]*nodegroup, error) {
```

Comment:
Need to add test cases for this method to verify that autodiscovery config is doing the right thing.